File: src-tauri/src/commands/note.rs
//!
//! This module contains Tauri commands for managing notes and tags within the note store.
//! It provides functionalities to create, retrieve, update, and delete notes and their
//! associated tags. The commands are designed to be invoked from the frontend,
//! allowing seamless interaction with the note-taking capabilities of the application.
//!
//! ## Overview
//!
//! - **Notes**: Functions to manage notes, including adding, retrieving, and deleting notes.
//! - **Tags**: Functions to manage note tags and their associations with notes.
//!
//! ## Usage
//!
//! The commands can be invoked from the frontend using Tauri's `invoke`
//! function. Each command is annotated with detailed documentation, including
//! parameters, return types, and examples of usage.
//!
//! ## Example
//!
//! ```js
//! import { invoke } from '@tauri-apps/api/core'
//! // Call from frontend to get all notes with a specific tag:
//! const notes = await invoke('get_notes', { tagId: 1 });
//! console.log(notes);
//! ```

use crate::db::{MainStore, Note, NoteTag};
use std::sync::Arc;
use tauri::{command, State};

use crate::error::{AppError, Result};

/// Add a new note
///
/// Creates a new note with the specified title, content, and tags.
///
/// # Arguments
/// - `state` - The state of the note store, automatically injected by Tauri
/// - `title` - The title of the note
/// - `content` - The content of the note
/// - `conversation_id` - Optional ID of the associated conversation
/// - `message_id` - Optional ID of the associated message
/// - `tags` - Comma-separated list of tags
///
/// # Returns
/// * `Result<(), String>` - Ok if successful or an error message
///
/// # Example
///
/// ```js
/// // Call from frontend:
/// import { invoke } from '@tauri-apps/api/core';
///
/// await invoke('add_note', {
///     title: 'My Note',
///     content: 'Note content',
///     conversationId: 1,
///     messageId: 2,
///     tags: 'rust,programming'
/// });
/// ```
#[command]
pub async fn add_note(
    state: State<'_, Arc<MainStore>>,
    title: String,
    content: String,
    conversation_id: Option<i64>,
    message_id: Option<i64>,
    tags: Vec<String>,
    metadata: Option<serde_json::Value>,
) -> Result<()> {
    let runtime = {
        let main_store = state.read()?;
        main_store.db_runtime().map_err(AppError::Db)?
    };
    MainStore::add_note_with_runtime(
        runtime,
        title,
        content,
        conversation_id,
        message_id,
        tags,
        metadata,
    )
    .await
    .map_err(AppError::Db)?;
    Ok(())
}

/// Get all tags
///
/// Retrieves a list of all tags from the note store.
///
/// # Arguments
/// - `state` - The state of the note store, automatically injected by Tauri
///
/// # Returns
/// * `Result<Vec<NoteTag>, String>` - A vector of tags or an error message
///
/// # Example
///
/// ```js
/// // Call from frontend:
/// import { invoke } from '@tauri-apps/api/core';
///
/// const tags = await invoke('get_tags');
/// console.log(tags);
/// ```
#[command]
pub async fn get_tags(state: State<'_, Arc<MainStore>>) -> Result<Vec<NoteTag>> {
    let runtime = {
        let main_store = state.read()?;
        main_store.db_runtime().map_err(AppError::Db)?
    };
    MainStore::get_tags_with_runtime(runtime)
        .await
        .map_err(AppError::Db)
}

/// Get notes by tag ID
///
/// Retrieves all notes associated with a specific tag.
///
/// # Arguments
/// - `state` - The state of the note store, automatically injected by Tauri
/// - `tag_id` - The ID of the tag to filter notes by
///
/// # Returns
/// * `Result<Vec<Note>, String>` - A vector of notes or an error message
///
/// # Example
///
/// ```js
/// // Call from frontend:
/// import { invoke } from '@tauri-apps/api/core';
///
/// const notes = await invoke('get_notes', { tagId: 1 });
/// console.log(notes);
/// ```
#[command]
pub async fn get_notes(state: State<'_, Arc<MainStore>>, tag_id: Option<i64>) -> Result<Vec<Note>> {
    let runtime = {
        let main_store = state.read()?;
        main_store.db_runtime().map_err(AppError::Db)?
    };
    MainStore::get_notes_with_runtime(runtime, tag_id)
        .await
        .map_err(AppError::Db)
}

/// Get a note by ID
///
/// Retrieves a specific note by its ID.
///
/// # Arguments
/// - `state` - The state of the note store, automatically injected by Tauri
/// - `id` - The ID of the note to retrieve
///
/// # Returns
/// * `Result<Note, String>` - The requested note or an error message
///
/// # Example
///
/// ```js
/// // Call from frontend:
/// import { invoke } from '@tauri-apps/api/core';
///
/// const note = await invoke('get_note', { id: 1 });
/// console.log(note);
/// ```
#[command]
pub async fn get_note(state: State<'_, Arc<MainStore>>, id: i64) -> Result<Note> {
    let runtime = {
        let main_store = state.read()?;
        main_store.db_runtime().map_err(AppError::Db)?
    };
    MainStore::get_note_with_runtime(runtime, id)
        .await
        .map_err(AppError::Db)
}

/// Delete a note
///
/// Removes a note and its tag associations from the store.
///
/// # Arguments
/// - `state` - The state of the note store, automatically injected by Tauri
/// - `id` - The ID of the note to delete
///
/// # Returns
/// * `Result<(), String>` - Ok if successful or an error message
///
/// # Example
///
/// ```js
/// // Call from frontend:
/// import { invoke } from '@tauri-apps/api/core';
///
/// await invoke('delete_note', { id: 1 });
/// console.log('Note deleted successfully');
/// ```
#[command]
pub async fn delete_note(state: State<'_, Arc<MainStore>>, id: i64) -> Result<()> {
    let runtime = {
        let main_store = state.read()?;
        main_store.db_runtime().map_err(AppError::Db)?
    };
    MainStore::delete_note_with_runtime(runtime, id)
        .await
        .map_err(AppError::Db)
}

/// Search notes
///
/// Searches for notes by keyword in their titles.
///
/// # Arguments
/// - `state` - The state of the note store, automatically injected by Tauri
/// - `kw` - The keyword to search for in note titles
///
/// # Returns
/// * `Result<Vec<Note>, String>` - Matching notes or an error message
///
/// # Example
///
/// ```js
/// // Call from frontend:
/// import { invoke } from '@tauri-apps/api/core';
///
/// const notes = await invoke('search_notes', { kw: 'rust' });
/// console.log('Found matching notes:', notes);
/// ```
#[command]
pub async fn search_notes(state: State<'_, Arc<MainStore>>, kw: String) -> Result<Vec<Note>> {
    let runtime = {
        let main_store = state.read()?;
        main_store.db_runtime().map_err(AppError::Db)?
    };
    MainStore::search_notes_with_runtime(runtime, kw)
        .await
        .map_err(AppError::Db)
}

File: src/stores/note.js
import { FrontendAppError, invokeWrapper } from '@/libs/tauri'
import { getCurrentWebviewWindow } from '@tauri-apps/api/webviewWindow'
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { sendSyncState } from '@/libs/sync'


export const useNoteStore = defineStore('note', () => {
  const label = getCurrentWebviewWindow().label;
  // Used to store all note tags. Each tag,
  // in addition to containing tag information,
  // also includes a list of all notes, structured as follows:
  // [{
  //     id: 1,
  //     name: '',
  //     note_count: 0,
  //     notes: []
  // }]
  const tags = ref([])
  /**
   * Gets a list of all note tags.
   */
  const getTagList = () => {
    invokeWrapper('get_tags').then(result => {
      if (!result) {
        tags.value = []
        return
      }
      tags.value = result.map(x => ({
        id: x.id,
        name: x.name,
        note_count: x.noteCount,
        notes: [],
      }))
      console.log(tags.value)
    }).catch(error => {
      if (error instanceof FrontendAppError) {
        console.error(`Error getting tag list: ${error.toFormattedString()}`, error.originalError);
      } else {
        console.error('Error getting tag list:', error);
      }
    })
  }

  /**
   * Creates a new note with the specified details and syncs the state across windows.
   * @param {string} title - The title of the note.
   * @param {string} content - The content of the note.
   * @param {number} [conversationId] - Optional ID of the associated conversation.
   * @param {number} [messageId] - Optional ID of the associated message.
   * @param {string} tags - Comma-separated list of tags.
   * @returns {Promise<void>} A promise that resolves when the note is created and state is synced.
   */
  const addNote = (title, content, conversationId, messageId, tags, metadata) => {
    tags = tags.replace('，', ',').split(',').map(x => x.trim()).filter(x => x !== '')
    console.log(metadata)
    return new Promise((resolve, reject) => invokeWrapper('add_note', { title, content, conversationId, messageId, tags, metadata })
      .then(() => {
        sendSyncState('note_update', label)
        resolve()
      }).catch(error => {
        if (error instanceof FrontendAppError) {
          console.error(`Error adding note: ${error.toFormattedString()}`, error.originalError);
        } else {
          console.error('Error adding note:', error);
        }
        reject(error)
      }))
  }

  /**
   * Gets a specific note by its ID.
   * @param {number} id - The ID of the note to retrieve.
   * @returns {Promise<Object>} A promise that resolves to the note object.
   */
  const getNote = async (id) => {
    return invokeWrapper('get_note', { id }).catch(error => {
      if (error instanceof FrontendAppError) {
        console.error(`Error getting note: ${error.toFormattedString()}`, error.originalError);
      } else {
        console.error('Error getting note:', error);
      }
      throw error;
    })
  }

  /**
   * Gets all notes associated with a specific tag.
   * @param {number} tagId - The ID of the tag to filter notes by.
   * @returns {Promise<Array>} A promise that resolves to an array of notes.
   */
  const getNotes = async (tagId) => {
    return invokeWrapper('get_notes', { tagId }).catch(error => {
      if (error instanceof FrontendAppError) {
        console.error(`Error getting notes: ${error.toFormattedString()}`, error.originalError);
      } else {
        console.error('Error getting notes:', error);
      }
      throw error;
    })
  }

  /**
   * Searches for notes based on a keyword.
   * @param {string} kw - The keyword to search for in note titles.
   * @returns {Promise<Array>} A promise that resolves to an array of matching notes.
   */
  const searchNotes = async (kw) => {
    return invokeWrapper('search_notes', { kw }).catch(error => {
      if (error instanceof FrontendAppError) {
        console.error(`Error searching notes: ${error.toFormattedString()}`, error.originalError);
      } else {
        console.error('Error searching notes:', error);
      }
      throw error;
    })
  }

  /**
   * Deletes a note by its ID.
   * @param {number} id - The ID of the note to delete.
   * @returns {Promise<void>} A promise that resolves when the note is deleted.
   */
  const deleteNote = async (id) => {
    return invokeWrapper('delete_note', { id }).catch(error => {
      if (error instanceof FrontendAppError) {
        console.error(`Error deleting note: ${error.toFormattedString()}`, error.originalError);
      } else {
        console.error('Error deleting note:', error);
      }
      throw error;
    })
  }

  return {
    windowLabel: label,
    tags,
    getTagList,
    addNote,
    getNote,
    getNotes,
    deleteNote,
    searchNotes
  }
})
